src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -504,15 +548,36 @@ bool ShenandoahOldHeuristics::should_start_gc() {
 
   if (_fragmentation_trigger) {
     ShenandoahHeap* heap = ShenandoahHeap::heap();
+    const size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
+
     size_t used = _old_generation->used();
     size_t used_regions_size = _old_generation->used_regions_size();
+
+    // used_regions includes humongous regions
     size_t used_regions = _old_generation->used_regions();
     assert(used_regions_size > used_regions, "Cannot have more used than used regions");
+
+    size_t first_active_region = heap->free_set()->first_old_region();
+    size_t last_active_region = heap->free_set()->last_old_region();
+    size_t span_of_active_regions = (last_active_region > first_active_region)? last_active_region + 1 - first_active_region: 0;
+
+    size_t first_old_region, last_old_region;
+    double density;
+    fragmentation_trigger_reason(density, first_old_region, last_old_region);
```

Comment:
I'm not sure which member fields you are recommending here.

ShenandoahFreeSet remembers first and last indexes of the OldCollector free set, but the range we are interested in here includes all retired old-gen regions in addition to the free regions.

If I'm misunderstanding your suggestion, please clarify.

Reply:
Not sure which member fields.  Please clarify.

Reply:
It looks like `fragmentation_trigger_reason` just copies the values of the member fields into its arguments.

Reply:
Thanks for clarifying.  I kind of like the encapsulating the update as is, because it makes more clear to the code maintainer what these fields represent, and how they got their values.  For performance, I'll add inline to the function definition.  Is that ok?

Reply:
Okay. Loooks like a typo in the log message: `densityr`. A comment describing what `density` is would also be welcome.

Reply:
Maybe a method `log_fragmentation_trigger_reason`?

Reply:
Thanks. Fixing the log message, and fixing this method name.